### PR TITLE
Introduce Client-Side Verilog Synthesis via Web Worker for Tauri Desktop

### LINKED ISSUE #1058: Feature: Introduce Client-Side Verilog Synthesis via Web Worker for Tauri Desktop
### Is your feature request related to a problem? Please describe

Yes. As part of my GSoC project (Project 7), I am bringing offline Verilog Synthesis to the Tauri Desktop application. However, implementing this presents two major technical problems:
1. **Main Thread Blocking**: Running the `@yowasp/yosys` WASM engine directly on the main thread is a heavy operation that freezes the application UI during circuit compilation.
2. **Repository Bloat**: The offline synthesis process requires generating custom worker bundles and downloading large `yosys.core*.wasm` binaries (~15MB). Committing these heavy artifacts directly to the repository is a bad version control practice that will bloat the codebase.

### Describe the solution you'd like

I will implement a robust architectural solution for the desktop app:
1. **Web Worker Integration**: Introduce a dedicated Web Worker to handle the Yosys WASM engine off the main thread, keeping the UI perfectly responsive during synthesis.
2. **Automated Build Pipeline**: Set up a clean, zero-friction developer workflow that automatically generates the worker bundles and handles the WASM binaries locally during `npm install` and `npm run dev`, without needing to track any of these heavy artifacts in version control.

### Describe alternatives you've considered

- **Server-Side API**: Relying on the existing Rails backend API. Rejected because the primary goal is to provide a fully offline capability for the desktop application.
- **Forking the whole yosys CI**: Rejected because it creates unnecessary repository bloat and leads to messy PRs.

### Is requested feature related to vue simulator?

Yes

### Additional context

This feature is a core deliverable for **GSoC 2026 Project 7: Client-Side Verilog Synthesis with Tauri**.

### Are you working on this?

Yes

### LINKED ISSUE #1059: refactor: migrate Tauri environment detection from internals to official isTauri API
### Description
Currently, the worker application relies on Tauri internals/window globals to detect whether it is running within the Tauri desktop environment. This approach is an anti-pattern and relies on undocumented internal behavior which could break in future Tauri updates.

This issue tracks the refactoring of environment detection to use the official `@tauri-apps/api/core` `isTauri()` method.

### Motivation
- **Best Practices:** Using official APIs rather than internal window globals ensures long-term stability.
- **Maintainability:** Makes platform-branching logic cleaner and easier to read.
- **GSoC 2026:** This fulfills a deliverable for the Client-Side Verilog Synthesis project to establish robust offline-routing infrastructure.

### Proposed Changes
- Remove legacy Tauri internal checks.
- Implement `isTauri()` for all desktop-specific branching (e.g., routing offline synthesis vs. server synthesis).

### Related PRs
- This refactor will be included and resolved alongside the upcoming Client-Side Verilog Synthesis PR.

### PR DESCRIPTION
**Pull request by Me-Priyank**
Fixes #1058 
Fixes #1059 

 

### Describe the changes you have made in this PR -

This PR introduces the foundational architecture for offline, client-side Verilog synthesis in the Tauri desktop application, ensuring the UI remains responsive and the repository remains clean. 
Key changes include:
1. **Non-Blocking Web Worker (`synthesisWorker.js`)**: Offloaded the `@yowasp/yosys` WASM execution and `yosys2digitaljs` conversion to a background thread to prevent UI freezing during synthesis.
2. **Asynchronous Orchestration (`clientSynthesis.js`)**: Implemented a `postMessage` protocol to manage synthesis requests, uniquely identifying them to support concurrent operations.
3. **Automated Worker Generation (`rollup.worker.config.mjs` & `package.json`)**: Created a standalone Rollup build process to handle Yosys BigInt dependencies (which esbuild fails on) and hooked it into the `postinstall` script. The worker bundle is now auto-generated locally for all contributors on `npm install`.
4. **Dynamic Asset Copying (`vite.config.ts`)**: Extended a custom Vite plugin to automatically copy the ~15MB of Yosys WASM binaries from `node_modules` to `public/assets/` during both `dev` server startup and production builds.
5. **Platform Routing (`Verilog2CV.js`)**: Replaced deprecated `IS_DESKTOP` logic with the official `@tauri-apps/api/core` `isTauri()` helper to intelligently route synthesis requests to either the Web Worker (desktop) or the Rails API (web).
6. **Clean Repository (`.gitignore`)**: Strictly excluded the generated worker bundle and WASM binaries from version control to prevent severe repository bloat.

### Workflow:

```mermaid
flowchart LR
    UI[Main Thread<br>Verilog2CV.js]
    Worker[Web Worker<br>synthesisWorker.js]
    WASM[(Yosys WASM)]
    Canvas[Circuit Canvas]

    UI -- "1. Send Verilog Code" --> Worker
    Worker -- "2. Execute Offline" --> WASM
    WASM -- "3. Return Netlist" --> Worker
    Worker -- "4. Send DigitalJS Circuit" --> UI
    UI -- "5. Auto-Layout" --> Canvas

    style UI fill:transparent,stroke:#2ea043,stroke-width:2px
    style Worker fill:transparent,stroke:#3b82f6,stroke-width:2px,stroke-dasharray: 5 5
    style WASM fill:transparent,stroke:#8b5cf6,stroke-width:2px
    style Canvas fill:transparent,stroke:#888,stroke-width:1px

```
### Screenrecording -

https://github.com/user-attachments/assets/831c0a0f-12af-4a0d-b503-5cf1007db151

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)
**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

### Explain your implementation approach:

- **What problem does your code solve?**

Running heavy WASM binaries (Yosys) on the browser's main thread blocks the event loop, causing the application UI to freeze completely. Furthermore, checking in the massive auto-generated worker file and WASM binaries would heavily bloat the Git repository and cause merge conflicts.

- **What alternative approaches did you consider?**

I considered compiling the worker via Vite's standard pipeline, but Vite's internal esbuild (v0.14) crashes when parsing the BigInt literals present in the `@yowasp/yosys` package. I also considered keeping the WASM files committed in the repo to simplify dev setup.

- **Why did you choose this specific implementation?**

I chose a dedicated Web Worker to ensure UI responsiveness. To solve the build issues without bloating the repo, I implemented a standalone Rollup config triggered via `postinstall`, combined with a Vite plugin to handle asset serving. This ensures every contributor automatically generates the exact files they need purely by running `npm install` and `npm run dev`, keeping the repo pristine.

- **What are the key functions/components and what do they do?**

  - `synthesisWorker.js`: The Web Worker script that receives code, loads the WASM, runs Yosys, and returns the parsed DigitalJS circuit.
  - `clientSynthesis.js`: The client-side manager that lazy-loads the worker, tracks request IDs, and resolves promises via `postMessage`.
  - `rollup.worker.config.mjs`: Bypasses esbuild to safely bundle the Yosys dependencies into a standalone worker file.
  - `copyYowaspAssets()` (in `vite.config.ts`): Hooks into the Vite dev server and build processes to ensure WASM binaries are available to the worker at runtime.
---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

## Summary by CodeRabbit

* **New Features**
  * Client-side Verilog synthesis (WASM) with progress updates
  * Automatic circuit auto-layout assigning stable x/y positions
  * Improved label positioning and alignment in subcircuits

* **Bug Fixes**
  * Non-overlapping, left-to-right ordered device placement; tolerates unknown or disconnected devices

* **Tests**
  * Added unit tests for synthesis and layout

* **Chores**
  * Build tooling and asset handling updates to support the synthesis worker and WASM assets

[Review Change Stack](https://app.coderabbit.ai/change-stack/CircuitVerse/cv-frontend-vue/pull/1060?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by Me-Priyank** [comment]
@coderabbitai review

**Comment by Me-Priyank** [comment]
@coderabbitai resolve

**Comment by Me-Priyank** [comment]
> Caution
> 
> Some comments are outside the diff and can’t be posted inline due to platform limitations.
> 
> ⚠️ Outside diff range comments (1)
> > package.json (1)> `66-66`: _⚠️ Potential issue_ | _🔴 Critical_ | _⚡ Quick win_
> > > **Downgrade `cross-env` to `^7.0.3` for Node 18 compatibility.**
> > > `cross-env` 10.x requires Node 20 or later. This repository uses Node 18, so installing dependencies will fail. Based on learnings, `cross-env` should be pinned to `^7.0.3` to maintain compatibility with Node 18 environments.
> > > 
> > > 🔧 Proposed fix
> > > ```diff
> > > -    "cross-env": "^10.1.0",
> > > +    "cross-env": "^7.0.3",
> > > ```
> 
> 🧹 Nitpick comments (1)
> > v1/src/simulator/src/Verilog2CV.js (1)> `288-343`: _🏗️ Heavy lift_
> > > **Drop the mirrored synthesis flow from `v1/` unless this path is still shipped.**
> > > This PR keeps duplicating active synthesis/error-handling logic into the legacy `v1/` tree, which doubles maintenance and drift risk for every follow-up fix. If `v1/` is no longer part of builds, these changes should stay in `src/simulator/src/` only.
> > > Based on learnings: in `CircuitVerse/cv-frontend-vue`, the `v1/` directory is legacy and not used in builds, so active development should stay in `src/simulator/src/`.
> 
> ℹ️ Review info

@coderabbitai  `cross-env@^10.1.0` already exists on main and was not modified by this PR, you can verify this line is absent from the diff. All CI workflows pin Node 22, so there's no compatibility issue. Out of scope for this PR.

and , `v1/` is actively built and tested, `scripts/multi-build.js` compiles both `src` and `v1`, and `vitest.workspace.ts` includes `createWorkspaceConfig("v1")` which runs v1 tests in CI. Removing the v1 mirror would break the v1 build. These changes must stay in both paths to keep them in sync.

**Comment by Me-Priyank** [comment]
@Nihal4777 @ThatDeparted2061 @092vk  ready for review

**Comment by Me-Priyank** [comment]
@092vk squashed all the commits

**Comment by Me-Priyank** [comment]
@Nihal4777 @ThatDeparted2061 @092vk PTAL

**Comment by Me-Priyank** [comment]
@Nihal4777 @ThatDeparted2061 @092vk ping...

**Comment by Me-Priyank** [comment]
**Note:** files like events.ts , data.ts etc... were churn from the earlier branch/conflict-resolution state. I have now kept them only as format-only changes to satisfy CI.(I first removed them from the PR, but the CI format check then failed because `oxfmt --check "**/*.ts"` runs repo-wide and reported those files.)

The actual work remains limited to introducing the worker only

File: src/simulator/src/Verilog2CV.js [hunk 1]
```diff
@@ -257,27 +288,49 @@ export function YosysJSON2CV(
...
+    if (isTauri()) {
+        setVerilogOutput('Compiling Verilog (offline, WASM)...', 'info')
+    } else {
+        setVerilogOutput('Compiling Verilog code...', 'info')
+    }
+
+    // Route synthesis based on platform:
+    //   Desktop (Tauri): client-side WASM via Web Worker (offline)
+    //   Web browser:     server-side Ruby gem via HTTP POST
+    var synthesisPromise = isTauri()
```
Comment (ThatDeparted2061) [inline src/simulator/src/Verilog2CV.js]: isTauri() called 3 times in generateVerilogCircuit so lets cache as const isDesktop = isTauri() once at function top.

File: .gitignore [hunk 2]
```diff
@@ -23,3 +23,10 @@ public/simulatorvue/*
 *.njsproj
 *.sln
 *.sw?
+
+# Generated synthesis worker bundle (auto-generated by `npm run build:worker` via postinstall)
```
Comment (ThatDeparted2061) [inline .gitignore:27]: postinstall runs Rollup on every npm install , i think if the Rollup build fails for any reason (network timeout during WASM package download, missing native deps in CI), the entire npm install aborts. so we shud add a fallback or move build:worker out of postinstall.

File: v1/src/simulator/src/Verilog2CV.js [hunk 3]
```diff
@@ -68,6 +71,37 @@ export function applyVerilogTheme(theme) {
...
+    if (typeof window !== 'undefined' && window.verilogTerminal) {
+        window.verilogTerminal.clearOutput()
+    } else {
+        const verilogOutputDiv = document.getElementById('verilogOutput')
+        if (verilogOutputDiv) {
+            verilogOutputDiv.innerHTML = ''
+        }
+    }
+}
+
```
Comment (ThatDeparted2061) [inline v1/src/simulator/src/Verilog2CV.js:104]: still using window.terminal + innerhtml XSS sink, an anti pattern, didnt we just remove that in 1055, either rebase on 1055 or update accordinly

File: src/simulator/src/synthesis/clientSynthesis.js
Comment (ThatDeparted2061) [inline src/simulator/src/synthesis/clientSynthesis.js:1]: I feel if if two synthesis requests are in flight and the worker errors, handleError for request A terminates the worker and nulls it out, but request B's handleMessage / handleError listeners stay permanently registered on a dead worker, they never fire and are never cleaned up.

File: v1/src/simulator/src/Verilog2CV.js [hunk 4]
```diff
@@ -258,17 +338,24 @@ export default function generateVerilogCircuit(
...
+                } else {
+                    showError('There is some issue with the code')
+                    error.json()
+                        .then((errorMessage) => {
+                            setVerilogOutput(errorMessage.message, 'error')
+                        })
+                        .catch(() => {
+                            // Response body wasn't valid JSON
+                        })
+                }
```
Comment (ThatDeparted2061) [inline v1/src/simulator/src/Verilog2CV.js]: error.json() rejection is silently swallowed, server-side non-JSON error responses (e.g. 502 nginx pages) silently drop the error, the user will get ghosted in that case. we aint his crush to let that happen to him. Add a .catch()

**Review by ThatDeparted2061 — CHANGES_REQUESTED** [review]

File: v1/src/simulator/src/Verilog2CV.js (hunk 4 above)
Comment (Me-Priyank) [reply v1/src/simulator/src/Verilog2CV.js]: Fixed... added a .catch() that surfaces a fallback message ('Server returned a non-JSON error response') so the user always gets terminal feedback.

File: src/simulator/src/synthesis/clientSynthesis.js
Comment (Me-Priyank) [reply src/simulator/src/synthesis/clientSynthesis.js:1]: Yes you're right , and I'm already aware about this . This is addressed in my Week 2 deliverable (Worker Lifecycle Management). The fix will be to maintain a pending-requests map keyed by `requestId`, so when the worker is terminated or errors, all pending promises are rejected, not just the one that triggered the error. I'll also move away from per-request `addEventListener`/`removeEventListener` to a single centralized `onmessage` dispatcher that routes by `requestId`, which eliminates stale listener registration entirely.

File: .gitignore (hunk 2 above)
Comment (Me-Priyank) [reply .gitignore:27]: Valid point.... although `build:worker` doesn't fetch anything, it just Rollup-bundles files from the already-installed `@yowasp/yosys` in `node_modules`, so network timeouts wouldn't affect it. But I agree postinstall should be fault-tolerant regardless.

So I've added a soft fallback instead so a failure warns but doesn't abort install:

```bash
\"postinstall\": \"husky install && (npm run build:worker || echo '[WARN] Worker build failed — run npm run build:worker manually')\"
```

This way install stays reliable while still generating the worker automatically

File: src/simulator/src/Verilog2CV.js (hunk 1 above)
Comment (Me-Priyank) [reply src/simulator/src/Verilog2CV.js]: done

File: v1/src/simulator/src/Verilog2CV.js (hunk 3 above)
Comment (Me-Priyank) [reply v1/src/simulator/src/Verilog2CV.js:104]: Once 1055 is merged, I'll rebase this PR on top of it so the terminal output flows through the store instead of raw DOM manipulation. That will also resolve the innerHTML XSS concern.

**Comment by Me-Priyank** [comment]
@ThatDeparted2061 @092vk @Nihal4777 PTAL
